### ISSUE TITLE
GenAPI should generate `throw null` method body for `DoesNotReturn` methods in ref.

### ISSUE THREAD

**Issue by safern** [dotnet/arcade#4820]
-  Is this issue blocking (no)
-  Is this issue causing unreasonable pain (no)

`GenAPI` should consider throwing null for methods that it's return type is `void` and it contains a `DoesNotReturnAttribute` to avoid compiler warnings, at the moment it would generate code that will generate compiler warnings because for the compiler the empty body is code that will return.

cc: @buyaa-n @eiriktsarpalis @stephentoub

**Comment by stephentoub** [comment]
cc: @jcouv

**Comment by jcouv** [comment]
Sounds good. Thanks

nit: I'd recommend generating `throw null` for all methods, not just `void`-returning ones or ones with `[DoesNotReturn]`. 
That's what the compiler does for generating ref assemblies :-)

**Comment by stephentoub** [comment]
> I'd recommend generating throw null for all methods, not just void-returning ones or ones with [DoesNotReturn]. That's what the compiler does for generating ref assemblies :-)

Doesn't that result in larger assemblies?

**Comment by jcouv** [comment]
> Doesn't that result in larger assemblies?

When the compiler does it internally, no (we're able to share, I don't remember the exact mechanism).
When doing it on source code, I think that we fold that too. That can be tested when the generated change is made (size increase should be small constant, not linear in number of affected methods).

**Comment by stephentoub** [comment]
> When the compiler does it internally, no

I see it emitting `ldnull; throw` for each method body.  You're saying it does something else?  Am I using /refout incorrectly?

**Comment by jcouv** [comment]
> I see it emitting ldnull and hrow for each method body. 

Is each method we pointing to a new IL blob? I expect that we're sharing a single blob in metadata for those two instructions. Aleksey also remembers something like that.

But I looked at the `refout` PRs and could not find test verifying this. I'll look around the code some more to confirm.

**Comment by stephentoub** [comment]
> Is each method we pointing to a new IL blob? I expect that we're sharing a single blob in metadata for those two instructions.

You're right.  That's happening for the refout.  It's not happening for folding methods in the actual source.

**Comment by ViktorHofer** [comment]
Closing as GenAPI got rewritten and now lives in dotnet/sdk: https://github.com/dotnet/sdk/tree/main/src/Compatibility
